bridge/bridge/deploy/local.py

```diff
@@ -0,0 +1,283 @@
+        ]
+
+        response = Response(resp.content, resp.status_code, headers)
+        return response
+
+    def run(self):
+        port = os.environ.get("PORT", 3000)
+        self.handle = threading.Thread(
+            target=self.app.run, kwargs={"host": "0.0.0.0", "port": port}
+        )
+        self.handle.start()
+
+
+def destroy_failed_conda(model_dir):
+    from mlflow.utils.conda import _get_conda_env_name  # type: ignore
+    from mlflow.pyfunc import _load_model_env  # type: ignore
+    from shutil import rmtree
+
+    env = _load_model_env(model_dir)
+    conda_env_path = f"{model_dir}/{env}"
+    logger.info(f"Conda env path: {conda_env_path}")
+    name = _get_conda_env_name(conda_env_path)
+    logger.info(f"Destroying Conda environment {name}.")
+    handle = Popen(
+        [
+            "conda",
+            "env",
+            "remove",
+            "-n",
+            name,
+        ],
+    )
+    handle.wait()
+    prefix = os.environ.get("CONDA_PREFIX", "/opt/conda")
+    conda_env_dir = f"{prefix}/envs/{name}"
+    logger.info(f"Conda Env Dir: {conda_env_dir}")
+    rmtree(conda_env_dir, ignore_errors=True)
+    logger.info(f"Done {handle.returncode}")
+
+
+def prepare_conda_environment(model_dir, returns):
+    from filelock import FileLock
+    from mlflow.utils.conda import get_or_create_conda_env  # type: ignore
+    from mlflow.pyfunc import _load_model_env  # type: ignore
+
+    env = _load_model_env(model_dir)
+    conda_env_path = f"{model_dir}/{env}"
+    lock = FileLock(CONDA_LOCK_FILE)
+    logger.info(f"Obtaining lock {CONDA_LOCK_FILE}.")
+    try:
+        with lock:
+            logger.info("Got lock.")
+            get_or_create_conda_env(conda_env_path)
+        returns[0] = 0
+    except Exception as e:
+        logger.exception(e)
+        returns[0] = 1
+
+
+class ModelDeployment:
+    def __init__(self, model_version: ModelVersion, model_uri: str, port: int):
+        self.model_version = model_version
+        self.model_uri = model_uri
+        self.serve_handle = None
+        self.port = port
+        self._install()
+
+    def _install(self):
+        # Install return code
+        self.returns = [None]
+        logger.info(f"Preparing Conda environment for {self.model_version}")
+        self.install_handle = threading.Thread(
+            target=prepare_conda_environment,
+            args=(
+                self.model_uri,
+                self.returns,
+            ),
+        )
+        self.install_handle.start()
+
+    def check(self):
+        # Check if currently installing
+        if self.install_handle.is_alive() or self.returns[0] is None:
+            logger.info(f"{self.model_version} installing...")
+            return
+
+        # Check if install complete and failed
+        if self.returns[0]:
+            logger.warning(f"{self.model_version} install failed.")
+            destroy_failed_conda(self.model_uri)
+            self._install()
```

Comment:
If the failure is a config error, wouldn't this trigger an infinite retry loop?

Reply:
Yes, but there were also spurious errors from Conda, so it needs to retry at least a few times. This can all definitely benefit from some refactoring at some point.

Reply:
Environment specification is probably the most finicky part of the MLmodel spec, so we should be on the lookout for this.

Reply:
I think we will need some product design work to decide how to handle this. Ideally it would mark a deployment as failed in the registry, and possibly roll back? 

How does Sagemaker handle this?